File: src/dose/dose_java.ss/src/com.saabgroup.safir.dob/src/com/saabgroup/safir/dob/EntityIterator.java
// -*- coding: utf-8 -*-
/******************************************************************************
*
* Copyright Saab AB, 2009-2013, 2021 (http://safirsdkcore.com)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
package com.saabgroup.safir.dob;

/**
 * An iterator that can be used to traverse entity instances.
 */
public class EntityIterator
    implements java.util.Iterator<EntityProxy>, AutoCloseable {
    /** Iterator constructor is internal to dose_java, since only
     * Connections (and SecondaryConnections) can instantiate them.
     */
    EntityIterator(int ctrl,
                   long typeId,
                   boolean includeSubclasses){
        m_state = new State();
        m_cleanable = ResourceHelper.register(this,m_state);

        m_state.ctrl = ctrl;

        int [] iteratorId = new int [1];
        boolean [] success = new boolean [1];

        Interface.EntityIteratorCreate(ctrl,
                                       typeId,
                                       includeSubclasses,
                                       iteratorId,
                                       m_state.end,
                                       success);
        m_state.iteratorId = iteratorId[0];

        if (!success[0])
        {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }

    /**
     * @see java.util.Iterator#hasNext()
     */
    @Override
    public boolean hasNext() {
        checkNotDisposed();
        if (m_state.moveToNext) {
            moveToNext();
            m_state.moveToNext = false;
        }

        return !m_state.end[0];
    }

    /**
     * @see java.util.Iterator#next()
     */
    @Override
    public EntityProxy next() {
        checkNotDisposed();
        if (m_state.moveToNext) {
            moveToNext();
        }
        //next time we will want to move on.
        m_state.moveToNext = true;

        //have we gone beyond the end of the instances?
        if (m_state.end[0]) {
            throw new java.util.NoSuchElementException("There are no more instances to iterate over! Have you used hasNext correctly?");
        }

        java.nio.ByteBuffer [] entityBlob = new java.nio.ByteBuffer[1];
        java.nio.ByteBuffer [] entityState = new java.nio.ByteBuffer[1];
        boolean [] success = new boolean [1];
        Interface.EntityIteratorDereference(m_state.ctrl,
                                            m_state.iteratorId,
                                            entityBlob,
                                            entityState,
                                            success);
        if (!success[0])
        {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }

        if (m_state.current != null) {
            throw new com.saabgroup.safir.dob.typesystem.SoftwareViolationException("Internal error in EntityIterator.next()! m_state.current was not null");
        }

        m_state.current = new EntityProxy(entityBlob[0].asReadOnlyBuffer(),entityState[0].asReadOnlyBuffer(),null,null,true,false);
        return m_state.current;
    }


    /**
     * @see java.util.Iterator#remove()
     */
    @Override
    public void remove() {
        throw new UnsupportedOperationException("EntityIterators do not support the remove operation");
    }

    private void moveToNext() {
        if (m_state.current != null) {
            m_state.current.dispose();
            m_state.current = null;
        }

        boolean [] success = new boolean [1];
        Interface.EntityIteratorIncrement(m_state.ctrl,
                                          m_state.iteratorId,
                                          m_state.end,
                                          success);
        if (!success[0])
        {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }

    /**
     * Release resources held by this iterator.
     */
    public void dispose() {
        m_state.dispose();
    }

    @Override
    public void close() {
        m_state.dispose();
    }

    private void checkNotDisposed()
    {
        if (m_state.disposed)
        {
            throw new com.saabgroup.safir.dob.typesystem.SoftwareViolationException("Attempt to use an EntityIterator that is disposed!");
        }
    }

    private static class State implements Runnable {
        public boolean disposed = false;
        public int ctrl;
        public int iteratorId;
        public boolean [] end = new boolean[1];
        public boolean moveToNext = false;
        public EntityProxy current = null;

        public void dispose() {
            if (!disposed) {
                disposed = true;

                if (current != null) {
                    current.dispose();
                }

                Interface.EntityIteratorDestroy(ctrl, iteratorId);
            }
        }

        /**
         * Clean up code
         */
        public void run() {
            try {
                if (!disposed) {
                    dispose();

                    com.saabgroup.safir.Logging.sendSystemLog
                        (com.saabgroup.safir.Logging.Severity.CRITICAL,
                         "Programming Error! An EntityIterator was not disposed correctly when cleanup was called!" +
                         "See the ConnectionBase#getEntityIterator(...).");
                }
            }
            catch (Exception exc) {
                com.saabgroup.safir.Logging.sendSystemLog
                    (com.saabgroup.safir.Logging.Severity.CRITICAL,
                     "EntityIterator.State.run: Caught exception: " + exc);
            }
        }
    }

    private final State m_state;
    private final java.lang.ref.Cleaner.Cleanable m_cleanable;
}

File: src/dose/dose_cpp.ss/src/EntityIterator.cpp
/******************************************************************************
*
* Copyright Saab AB, 2008-2013 (http://safir.sourceforge.net)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
#include <Safir/Dob/EntityIterator.h>

#include <Safir/Dob/Typesystem/LibraryExceptions.h>
#include <Safir/Dob/Internal/Interface.h>
#include "EntityProxyImpl.h"

namespace Safir
{
namespace Dob
{
    EntityIterator::EntityIterator():
        m_ctrl(-1),
        m_iteratorId(-1)
    {

    }

    EntityIterator::EntityIterator(const long ctrl,
                                   const Typesystem::TypeId typeId,
                                   const bool includeSubclasses):
        m_ctrl(ctrl),
        m_iteratorId(-1)
    {
        bool success;
        bool end;
        DoseC_EntityIteratorCreate(m_ctrl, typeId, includeSubclasses, m_iteratorId, end, success);

        if (!success)
        {
            Typesystem::LibraryExceptions::Instance().Throw();
        }

        if (end)
        {
            *this = EntityIterator();
        }
    }


    EntityIterator::EntityIterator(const EntityIterator & other):
        m_ctrl(other.m_ctrl),
        m_iteratorId(-1)
    {
        if (other.m_iteratorId != -1)
        {
            bool success;
            DoseC_EntityIteratorCopy(m_ctrl, other.m_iteratorId, m_iteratorId, success);

            if (!success)
            {
                Typesystem::LibraryExceptions::Instance().Throw();
            }
        }
    }

    EntityIterator::~EntityIterator()
    {
        if (m_iteratorId != -1)
        {
            DoseC_EntityIteratorDestroy(m_ctrl, m_iteratorId);
            m_iteratorId = -1;
        }
    }

    EntityIterator & EntityIterator::operator=(const EntityIterator& other)
    {
        m_dereferenced.reset();

        if (m_iteratorId != -1)
        {
            DoseC_EntityIteratorDestroy(m_ctrl, m_iteratorId);
            m_iteratorId = -1;
        }
        m_ctrl = other.m_ctrl;
        if (other.m_iteratorId != -1)
        {
            bool success;
            DoseC_EntityIteratorCopy(m_ctrl, other.m_iteratorId, m_iteratorId, success);

            if (!success)
            {
                Typesystem::LibraryExceptions::Instance().Throw();
            }
        }
        return *this;
    }


    const EntityProxy& EntityIterator::dereference() const
    {
        if (m_iteratorId == -1)
        {
            throw Typesystem::SoftwareViolationException(L"This iterator appears to be at 'end', so dont try to dereference it!", __WFILE__,__LINE__);
        }

        if (m_dereferenced == NULL)
        {
            bool success;
            const char* entityBlob;
            const char* entityState;
            DoseC_EntityIteratorDereference(m_ctrl, m_iteratorId, entityBlob, entityState, success);

            if (!success)
            {
                Typesystem::LibraryExceptions::Instance().Throw();
            }
            m_dereferenced.reset(new EntityProxy
                (new Internal::EntityProxyImpl(entityBlob,
                                               entityState,
                                               NULL,
                                               NULL,
                                               true,  //addReference
                                               false))); //timestampDiff
        }

        return *m_dereferenced;
    }

    void EntityIterator::increment()
    {
        if (m_iteratorId == -1)
        {
            throw Typesystem::SoftwareViolationException(L"Cannot increment an 'end' iterator", __WFILE__,__LINE__);
        }

        m_dereferenced.reset();

        bool success;
        bool end;
        DoseC_EntityIteratorIncrement(m_ctrl, m_iteratorId, end, success);

        if (!success)
        {
            Typesystem::LibraryExceptions::Instance().Throw();
        }

        if (end)
        {
            *this = EntityIterator();
        }
    }

    bool EntityIterator::equal(const EntityIterator& other) const
    {
        if (m_iteratorId == -1 && other.m_iteratorId == -1)
        {
            return true;
        }
        else if (m_iteratorId == -1 || other.m_iteratorId == -1)
        {
            return false;
        }

        if (m_ctrl != other.m_ctrl)
        {
            throw Typesystem::SoftwareViolationException(L"Cannot compare iterators from different connections!", __WFILE__,__LINE__);
        }

        bool success;
        bool equal;
        DoseC_EntityIteratorEqual(m_ctrl, m_iteratorId, other.m_iteratorId, equal, success);

        if (!success)
        {
            Typesystem::LibraryExceptions::Instance().Throw();
        }
        return equal;
    }

}
}
